Restructure Action handling & fix relation handling with multiple actions

jsangmeister opened the pull request:
fixes #321

normanjaeckel commented on dev/requirements_development.txt:
```diff
@@ -7,5 +7,6 @@ pytest==6.0.2
 pytest-cov==2.10.1
 pyyaml==5.3.1
 pip-check==2.6
+autoflake==1.4
```
What does this tool do and where is it used?

jsangmeister replied:
It removes unneeded import which are reported by flake. It was really annoying to remove them all by hand... It's used in the Makefile

normanjaeckel commented on openslides_backend/action/util/typing.py:
```diff
@@ -20,14 +18,3 @@
 ActionError = Any
-
-
-class Action(Protocol):  # pragma: no cover
```
You have to re-add this protocol (and also rename it). It is used in http/views.py.

jsangmeister replied:
I don't see why it's needed there, it was only used for one cast

normanjaeckel commented on openslides_backend/action/action.py:
```diff
@@ -1,139 +1,408 @@
 
-        # Parse actions and creates events
-        write_request_element, results = self.parse_actions(payload)
+    def update_instance(self, instance: Dict[str, Any]) -> Dict[str, Any]:
```
I think this method is not useful here. Where do we need it? The code can also be in base_update_instance and if some mixins want to create generic code they can define `special_update_instance` method.

jsangmeister replied:
Yes, I generally agree, but all three generic actions (create, update, delete) use such a method and it would therefore be defined in all three files. I found that unnecessary and moved it one level up. We could write `return self.update_instance(instance)` in the `base_update_instance` to symbolize the usage, if you want. Or make a better suggestion, I'm not so sure about this either.

normanjaeckel replied:
Ok, I follow your suggestion in #358

jsangmeister commented:
Should be done now and simplify the general handling and implementation of future changes.

jsangmeister commented:
Since the PR got pretty big, here a summary of the changes:
- Relation handling was moved to a `RelationManager` which is single-handedly responsible for handling all relation correctly. It uses the old `RelationsHandler` (now `SingleRelationHandler`) as a helper class and takes care fo repeated updates of the same field. It is instanciated once for each call to the backend.
- Restructured the `action` folder (this is probably where most of the changes come from). To not mix the general action classes with the specific actions, the latter were moved into the folder `actions`. I'm not too happy with the subpath `action/actions/.../` but didn't have a better name for it. All "base" classes for the actions were moved into appropriate subfolders and sometimes renamed.
- Added `autoflake` ro remove unneeded imports.
- Restructured the `perform` method of the actions. It is now clearer than before (I hope), what is done when and where one would need to intervene for specific changes. It also takes care of all relation handling now since this is shared by all actions.
- As a first step towards the transaction model described in #325, each payload item is now validated alone and not together with the other payload. This lead to the implementation fo the `SingularPayloadMixin` to still be able to have actions which take exactly one item.
- Extended the usage of the Model meta class. This enables direct field access, so one does not have to iterate all fields each time a specific field is accessed. Also implemented this as a prefix map to unify the handling with structured fields (which have unique prefixes). Replaced most tuples `(field_name, field)` with just the field; the name can be accessed via `field.own_field_name`.

jsangmeister commented:
Merging this since multiple changes depend on this. If you have any remarks, you can still add them

jsangmeister commented:
Link #325

normanjaeckel commented on openslides_backend/models/base.py:
```diff
@@ -21,11 +22,18 @@ def __new__(metaclass, class_name, class_parents, class_attributes):  # type: ig
             metaclass, class_name, class_parents, class_attributes
         )
         if class_name != "Model":
+            new_class.field_prefix_map = {}
```
Where do we use this new field_prefix_map? I can not find it in the code.

Also: Why do we want to save the template fields twice, once with the pythonic attribute name (e. g. `group__ids` for `User` and once with a shortend name (e. g. `group_`)?

jsangmeister replied:
I would not call it magic, but yes, thank you for the docs :D my bad for that.
